File: src/main/scala/net/ladstatt/javacv/fx/WebcamFXController.scala
package net.ladstatt.javacv.fx

import java.net.URL
import java.util.ResourceBundle

import javafx.beans.property.{SimpleBooleanProperty, SimpleObjectProperty}
import javafx.beans.value.{ChangeListener, ObservableValue}
import javafx.fxml.{FXML, Initializable}
import javafx.scene.image._
import net.ladstatt.javacv.PoorMansTimer
import org.bytedeco.javacv.{Frame, OpenCVFrameConverter, OpenCVFrameGrabber}
import zio._

abstract class WebcamFXController extends Initializable {

  val zioRuntimeProperty = new SimpleObjectProperty[Runtime[zio.ZEnv]]()

  @FXML var videoView: ImageView = _

  def setZioRuntime(runtime: Runtime[zio.ZEnv]): Unit = zioRuntimeProperty.set(runtime)

  def getZioRuntime(): Runtime[zio.ZEnv] = zioRuntimeProperty.get()

  protected def updateView(frame: Frame): Unit

  val javaCVConv = new OpenCVFrameConverter.ToMat

  /** controls if application closes */
  val cameraActiveProperty = new SimpleBooleanProperty(true)

  def setCameraActive(isActive: Boolean): Unit = cameraActiveProperty.set(isActive)

  def getCameraActive: Boolean = cameraActiveProperty.get

  def shutdown(): Unit = setCameraActive(false)

  val frameGrabber: OpenCVFrameGrabber = {
    val grabber = new OpenCVFrameGrabber(0)
    grabber.start()
    grabber
  }

  /*
    val videoObservable: Observable[Frame] =
      Observable[Frame](o => {
        new Thread(
          () => {
            while (getCameraActive) {
              Try {
                frameGrabber.grab()
              } match {
                case Success(m) => o.onNext(m)
                case Failure(e) => o.onError(e)
              }
            }
            frameGrabber.release()
            o.onCompleted()
          }).start()
      })

    */
  def setVideoView(mat: Frame): Unit = {
    PoorMansTimer.time(updateView(mat))
  }

  val loop = for {
    frame <- ZIO.succeed(frameGrabber.grab())
    _ <- ZIO.succeed(setVideoView(frame))
  } yield ()


  override def initialize(url: URL, resourceBundle: ResourceBundle): Unit = {
    zioRuntimeProperty.addListener(new ChangeListener[Runtime[ZEnv]] {
      override def changed(observableValue: ObservableValue[_ <: Runtime[zio.ZEnv]], t: Runtime[zio.ZEnv], t1: Runtime[zio.ZEnv]): Unit = {
        Option(t1) match {
          case Some(rt) => new Thread(() => rt.unsafeRun(loop.forever)).start()
          case None =>
        }
      }
    }

    )
    // videoSource.left
    //    getZioRuntime().unsafeRun(Task(println("Hello World!")))
    /*
    videoObservable.subscribe(
      (frame: Frame) => setVideoView(frame)
      , t => t.printStackTrace()
      , () => println("Videostream stopped..."))
    ( )

   */
  }

}

File: src/main/java/net/ladstatt/javacv/fx/WebcamFXController.java
package net.ladstatt.javacv.fx;


import javafx.application.Platform;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.image.*;
import org.bytedeco.javacv.Frame;
import org.bytedeco.javacv.FrameGrabber;
import org.bytedeco.javacv.OpenCVFrameConverter;
import org.bytedeco.javacv.OpenCVFrameGrabber;
import org.bytedeco.opencv.global.opencv_imgproc;
import org.bytedeco.opencv.opencv_core.Mat;

import java.net.URL;
import java.nio.ByteBuffer;
import java.util.ResourceBundle;

import static org.opencv.imgproc.Imgproc.COLOR_BGR2BGRA;

public class WebcamFXController implements Initializable {

    @FXML
    ImageView videoView;

    Mat javaCVMat = new Mat();

    /**
     * create buffer only once saves much time!
     */
    WritablePixelFormat<ByteBuffer> formatByte = PixelFormat.getByteBgraPreInstance();

    OpenCVFrameConverter<Mat> javaCVConv = new OpenCVFrameConverter.ToMat();

    /**
     * controls if application closes
     */
    SimpleBooleanProperty cameraActiveProperty = new SimpleBooleanProperty(true);

    OpenCVFrameGrabber frameGrabber = new OpenCVFrameGrabber(0);

    ByteBuffer buffer;

    protected void updateView(Frame frame) {
        int w = frame.imageWidth;
        int h = frame.imageHeight;

        Mat mat = javaCVConv.convert(frame);
        opencv_imgproc.cvtColor(mat, javaCVMat, COLOR_BGR2BGRA);
        if (buffer == null) {
            buffer = javaCVMat.createBuffer();
        }

        PixelBuffer<ByteBuffer> pb = new PixelBuffer<ByteBuffer>(w, h, buffer, formatByte);
        final WritableImage wi = new WritableImage(pb);
        Platform.runLater(() -> videoView.setImage(wi));

    }


    public void setCameraActive(Boolean isActive) {
        cameraActiveProperty.set(isActive);
    }

    public Boolean getCameraActive() {
        return cameraActiveProperty.get();
    }

    public void shutdown() {
        setCameraActive(false);
    }

    void setVideoView(Frame mat) {
        updateView(mat);
    }

    public void initialize(URL url, ResourceBundle resourceBundle) {
        try {
            frameGrabber.start();
        } catch (FrameGrabber.Exception e) {
            e.printStackTrace();
        }
        new Thread(() -> {
            while (getCameraActive()) {
                try {
                    Frame frame = frameGrabber.grab();
                    setVideoView(frame);
                } catch (FrameGrabber.Exception e) {
                    e.printStackTrace();
                }
            }
            try {
                frameGrabber.release();
            } catch (FrameGrabber.Exception e) {
                e.printStackTrace();
            }
        }).start();
    }

}
